### ISSUE TITLE
Size of ULID is 20 bytes not 16 as expected

### ISSUE THREAD

**Issue by etylermoss** [pksunkara/pgx_ulid#37]
I'm interfacing with the JVM / JDBC, and when I call `ResultSet.getBytes(col)`, I am getting an exception because the column is not 16 bytes.

I confirmed this is the case by executing the below command in postgres:

```sql
select gen_ulid(), pg_column_size(gen_ulid()), gen_random_uuid(), pg_column_size(gen_random_uuid())
```
| `pg_column_size(gen_ulid())` | `pg_column_size(gen_random_uuid())` |
|--------|--------|
| 20 | 16 | 

Any idea how to resolve this?

**Comment by pksunkara** [comment]
That is weird. I have u128 all over in my code and nowhere do I store it as 20 Bytes.

**Comment by etylermoss** [comment]
This may be an issue related to JDBC or casting operator, I just debugged the `ResultSet.getBytes(col)`, and the resulting `ByteArray` is 26 bytes long corresponding to the ULID string. Although I don't think that explains why `pg_column_size` is 20 and not 16.

**Comment by pksunkara** [comment]
The 26 bytes in the result is correct. But yeah, I'm unsure of what `pg_column_size` is supposed to be.

**Comment by etylermoss** [comment]
Perhaps `pg_column_size` includes padding / alignment?

Anyway, I have found the source of the issue, the custom type needs to be registered by the JDBC driver: https://github.com/pgjdbc/pgjdbc/issues/2554.

I will update once I have a solution for anyone else who encounters this.

**Comment by etylermoss** [comment]
@pksunkara can you think of anything that may need to be done on the database to support transfer? I'm new to PG so unsure exactly what I'm looking for - but I'm debugging the JDBC driver and it looks like the raw data being received is already serialized when it comes in. Do you know if the driver / connection needs to request the data be sent in a binary format; or if such a configuration is done for the type itself on database side?

**Comment by pksunkara** [comment]
> such a configuration is done for the type itself on database side

Yeah, it's a feature on extension side. We store the data in a space efficient way, but we show the data to the user in a readable way.

Unfortunately, I don't know what else I can do to help you debug. #27 might be related though. You can try asking around in https://discord.gg/PMrpdJsqcJ which is the framework for Postgres Extensions I use.

**Comment by arjunyel** [comment]
I created my ulids client using the JS library https://www.npmjs.com/package/ulidx, my `pg_column_size` is coming back as 17

**Comment by elektro-wolle** [comment]
It's caused by an internal representation in postgresql as [varlena](https://github.com/pgcentralfoundation/pgrx/blob/develop/docs/src/pg-internal/varlena.md): 4 bytes for the length and 16 bytes for the data. 
Root cause: in the pgrx-sql-entity-graph crate, the sql for the type is rendered in [`src/postgres_type/entity.rs`](https://docs.rs/pgrx-sql-entity-graph/latest/src/pgrx_sql_entity_graph/postgres_type/entity.rs.html#159) as 

```sql
CREATE TYPE ulid (
	INTERNALLENGTH = variable,
	INPUT = ulid_in, /* ulid::ulid_in */
	OUTPUT = ulid_out, /* ulid::ulid_out */
	STORAGE = extended
);
```

But it should be

```sql
CREATE TYPE ulid (
        INTERNALLENGTH = 16,
        INPUT = ulid_in, /* ulid::ulid_in */
        OUTPUT = ulid_out, /* ulid::ulid_out */
        STORAGE = plain
);
```

**Comment by pksunkara** [comment]
@elektro-wolle Thank you. Really appreciate it.

Looks related to https://github.com/pgcentralfoundation/pgrx/issues/143. @workingjubilee Should `PostgresType` support a `length` derive property to support this?

**Comment by k0nserv** [comment]
Hey 👋🏼 

I took some inspiration from this project when building [plid](https://github.com/k0nserv/plid/), thanks! I also noticed that the size ended up as 20 bytes and worked to get it down to 16 bytes in `plid`. Unfortunately this required opting out of `PostgresType` and other derives and dropping down to a lower level. Feel free to take inspiration from my code
